### ISSUE TITLE
Consider hyphenating action names when using LowercaseUrls

### ISSUE THREAD

**Issue by glen-84** [aspnet/Routing#186]
`SomeLongActionName` in the Home controller will be converted to `/home/somelongactionname`, which is difficult to read. `/home/some-long-action-name` is better, IMHO.

This can be done with Regex ([example](http://stackoverflow.com/a/22650308/221528)), but I'm not sure what the performance is like.

If you decide that this should not be the default, then perhaps you could add a setting (LowercaseUrlWordDelimiter?) with the default of an empty string. This would also allow people to use an underscore, for example.

**Comment by danroth27** [comment]
@rynowak Could this be done by reformatting all the action names using the Application Model?

**Comment by rynowak** [comment]
The question is whether or not you want to canonicalize the action-names or canonicalize the URLs being generated..

For instance, one might implement this feature by transforming all of the URLs going in and out of the router, by converting from kebab-case to PascalCase and vice-versa. This of course also means that your transformation algorithm has to operate *safely* on all path-tokens of the URL.

If you did this with application-model then you're going to affect the inputs at callsites to `Url.Action(...)`, **or** you'd need to implement a custom `UrlHelper` as well.

We might want to consider replacing the "lowercase URLs" feature with a pluggable strategy for canonicalization. Then you're feature to use whatever strategy you want, as long as it can roundtrip.

**Comment by glen-84** [comment]
I don't think that the inputs (controller/action names) to `Url.Action` should change, it should only affect the generated URLs (and be understood by the router).

**Comment by khalidabuhakmeh** [comment]
> We might want to consider replacing the "lowercase URLs" feature with a pluggable strategy for canonicalization. Then you're free to use whatever strategy you want, as long as it can roundtrip.

I like this idea, as I've seen some sites that use `+` to seperate their urls.

:+1:

**Comment by OlsonDev** [comment]
+1; Any update on this? Is there a workaround one can use today?

**Comment by glen-84** [comment]
@OlsonDev I haven't worked with ASP.NET for a few months, but I think that you just have to use explicit routes like this:

```c#
[HttpGet("add-phone-number")]
public IActionResult AddPhoneNumber()
{
    return View();
}
```

**Comment by OlsonDev** [comment]
@glen-84 That's what I have now; it's tedious to maintain. I would love this to be conventions-based. I made my own custom conventions thing around the MVC3 era and I'd be fine with that now but I'm unclear as to where tap into.

**Comment by rynowak** [comment]
You can use conventions to customize routing - it should be pretty straightforward. https://github.com/aspnet/Entropy/tree/dev/samples/Mvc.CustomRoutingConvention

**Comment by OlsonDev** [comment]
@rynowak Thanks -- but I ran into a bit of an issue. For sake of example, let's say I have `AwesomeUserController` with actions `Index`, `AddPhoneNumber`, etc. I want these routes to work:

```
/awesome-user -- lists all users
/awesome-user/index -- output /awesome-user, inform browser/client canonical URL is /awesome-user
/awesome-user/index/1 -- lists AwesomeUserId 1's info
/awesome-user/add-phone-number -- should return invalid request response
/awesome-user/add-phone-number/1 -- expects POST data [FromBody]
```

So, on the Controller I add a template `awesome-user` and on the actions I add `index/{id?}` and `add-phone-number/{id?}`. Now `/awesome-user` doesn't work because `index` is required.

This example extends to `HomeController` having the template `home`; `/` & `/home` wouldn't work anymore and I'd have to enter `/home/index`. I can get `/` to work if I add two templates to HomeController (`"home"` and `""`) and specifically **don't** add `Index()`'s template. Then, of course `/home/index/1` wouldn't work.

The problem is `Microsoft.AspNet.Mvc.ApplicationModels.ActionModel` only has one `AttributeRouteModel` and no way to make `[action]` optional; telling it to use `{action}`, `{action?}`, or `[action?]` causes it to throw errors:
```
Error: The attribute route 'awesome-user/{action}/{id?}' cannot contain a parameter named '{action}'. Use '[action]' in the route template to insert the value 'Index'.
... or ...
Error: The attribute route 'awesome-user/{action?}/{id?}' cannot contain a parameter named '{action}'. Use '[action]' in the route template to insert the value 'Index'.
... or ...
Error: While processing template 'awesome-user/[action?]/{id?}', a replacement value for the token 'action?' could not be found. Available tokens: 'action, controller'.
```

I'm at a loss as to how to make this convention support all those use-cases. Thoughts?

**Comment by OlsonDev** [comment]
Just noticed https://github.com/aspnet/Mvc/issues/4043 -- and based on what I read, solved my issue. However, now my solution is contigent on what you guys plan on doing. This is what I came up with:

```c#
public class HyphenatedRoutingConvention : IApplicationModelConvention {
    public string DefaultController { get; set; }
    public string DefaultAction { get; set; }

    public HyphenatedRoutingConvention(string defaultController = "home", string defaultAction = "index") {
        DefaultController = defaultController.ToLower();
        DefaultAction = defaultAction.ToLower();
    }

    public void Apply(ApplicationModel application) {
        foreach (var controller in application.Controllers) {
            if (controller.AttributeRoutes.Count != 0) continue;

            var controllerTmpl = Slugify(controller.ControllerName);
            controller.AttributeRoutes.Add(new AttributeRouteModel { Template = controllerTmpl });
            if (controllerTmpl == DefaultController) {
                controller.AttributeRoutes.Add(new AttributeRouteModel { Template = "" });
            }

            var actionsToAdd = new List<ActionModel>();
            foreach (var action in controller.Actions) {
                if (action.AttributeRouteModel != null) continue;
                var actionSlug = Slugify(action.ActionName);
                var actionTmpl = $"{actionSlug}/{{id?}}";
                if (actionSlug == DefaultAction) {
                    var defaultActionModel = new ActionModel(action) { AttributeRouteModel = new AttributeRouteModel { Template = "" } };
                    actionsToAdd.Add(defaultActionModel);
                }
                action.AttributeRouteModel = new AttributeRouteModel { Template = actionTmpl };
            }
            // flaky; contingent on resolution of https://github.com/aspnet/Mvc/issues/4043
            foreach (var action in actionsToAdd) {
                controller.Actions.Add(action);
            }
        }
    }

    private string Slugify(string pascalString) {
        return Regex.Replace(pascalString, "([a-z])([A-Z])", "$1-$2").ToLower();
    }
}
```

**Comment by rynowak** [comment]
If you're trying to mimic the behavior of `{controller}/{action}/{id?}` but with slugified names, I don't think it's as hard as what you're doing. You should really only have to double up on the first action, not on the controller.

Similar to your original table - but updating it for "home":
```
`/` -- goes to HomeController.Index()
`/home` -- goes to HomeController.Index()
`/home/index` -- goes to HomeController.Index()
`/home/other-action` -- goes to HomeController.OtherAction()
```

```
public class HomeController
{
    public IActionResult Index() { ... }
    public IActionResult OtherAction() { ... }
}
```

```
var actionSlug = Slugify(action.ActionName);
if (actionSlug == DefaultAction) {

    // This is the default action - add "" so it will be combined with [controller]
    actionsToAdd.Add(new ActionModel(action)
    { 
        AttributeRouteModel = new AttributeRouteModel { Template = "" }
    });

    // If this is the default action on the default controller - add "/" so it will match the root
    if (controllerTmpl == DefaultController) {
        actionsToAdd.Add(new ActionModel(action)
        { 
            AttributeRouteModel = new AttributeRouteModel { Template = "/" }
        });
    }
}
```

This works because `/` and `~/` override whatever comes before - so the `[controller]` prefix is dropped.

@Eilon @danroth27 - thoughts?

**Comment by OlsonDev** [comment]
Indeed, I'm trying to mimic that behavior. I got that to work but I wouldn't describe that as "easier" per se (same # SLOC, higher cyclomatic complexity :stuck_out_tongue:) -- more efficient, sure (I'm guessing). Good to know @ `/` and `~/`.

Am I right in regards to what I'll need to change after https://github.com/aspnet/Mvc/issues/4043 lands?

**Comment by rynowak** [comment]
If you do it my way there won't be any change :+1: 

> more efficient, sure (I'm guessing). Good to know @ / and ~/.

Regard of how you manipulate the model to do this, I'd expect the same route table to be generated.

**Comment by OlsonDev** [comment]
> If you do it my way there won't be any change :+1:

With your way, I think
```c#
var controllerTmpl = Slugify(controller.ControllerName);
controller.AttributeRoutes.Add(new AttributeRouteModel { Template = controllerTmpl });
```
would become
```c#
var controllerTmpl = Slugify(controller.ControllerName);
controller.AttributeRouteModel = new AttributeRouteModel { Template = controllerTmpl };
```

> I'd expect the same route table to be generated

I was guessing startup efficiency would be better; less work to generate the same table. Either way, probably an insignificant difference.

Thanks for all the help!

**Comment by Eilon** [comment]
No plans to implement this feature.

**Comment by rynowak** [comment]
Reopening this for consideration based on the current routing work. It would be great to have an extensibility point that lets you transform incoming and outgoing parameter values.

**Comment by mkArtakMSFT** [comment]
@glennc, thoughts?

**Comment by JamesNK** [comment]
Some thoughts on this:
1. This will need to be able to round-trip. The link generator will need to encode the value with dashes and the matcher will then need to make that back up on route match. A simple lowercase doesn't have this problem because matching is not case sensitive.
2. For good performance this will need to be baked into the matcher. When we get an incoming URL we don't want to convert areas/controllers/actions to be snake-case on the fly as we compare against them.

    Newtonsoft.Json has a similar feature and a property name is resolved once to its camel case or snake case equivalent when a contract is resolved. After that it is a simple string comparison against the resolved name on deserialization. No decoding is needed.

    The strategy will either need to add additional match states, e.g. `EditAccountController` can be matched by both `editaccount` or `edit-account`, or replace the default match state.
3. What do we want to format? Area, controller, action seem obvious, but what about a "productName" route parameter? `.Url(new { action = "Products", productName = "ContosoHeadlightFluid" })` = `/products/contoso-headlight-fluid`. That could get problematic fast when decoding.
4. Where would the option live? It shouldn't be LinkOptions because it effects matching as well.

**Comment by rynowak** [comment]
I don't think we'd want to implement something regarding snake-case specifically.

more like: we'd want to allow user code to plug in at the right places to take control. 

We've gotten several requests like this (all different), and I'm not interested in us actually implementing different strategies ourselves. The 'official' MVC style looks like `/Home/Index/15` and we're not trying to recommend something different.

-- 
I think this is a good summary of the issues and requirements overall.

Responding to your points...

1. To me this also implies that user code has to be involved in how MVC 'flattens' routes. This is the only place in the system that parameters are turned into non-parameters (other than URL generation). If the contract is really simple (`Func<(string key, string value), result>`) then this is easy. 

2.
> When we get an incoming URL we don't want to convert areas/controllers/actions to be kebab-case on the fly as we compare against them.

Maybe we're saying the same thing, but I think this needs to be done during the phase where we create the endpoints. 

> The strategy will either need to add additional match states, e.g. EditAccountController can be matched by both editaccount or edit-account, or replace the default match state.

I think it should replace the default casing. This is a stylistic choice for building a URL-space, and I think you should only do that with conviction. 

3. 
>What do we want to format? Area, controller, action seem obvious, but what about a "productName" route parameter?

I think user-code needs to decide. 


> In Newtonsoft.Json there is no logic anywhere to decode a camelcase or kebabcase value back to what it was. IMO these encodings are one way. For example is the camelcase urlCase decoded as UrlCase or URLCase? (or even was it already camel when it was passed in...)

Yeah I'm still not sold on the idea that we need to provide de-coding. 


4. 
> Where would the option live? It shouldn't be LinkOptions because it effects matching as well.

I like the idea that you could plug this in via the same mechanisms that support constraints today. This is problematic though, because that means we need to resolve services during the building on endpoints. Maybe the normalization isn't reflected in the endpoint? Ugh. I don't like that either but it seems the least-bad. 

I think defining how parameters get transformed has to be part of the route/template/endpoint definition rather than global. we're envisioning a system where lots of things contribute to routing, so we shouldn't build features that are *only* global.

I can also imagine how you could use `EndpointBuilder` to add things to endpoints you didn't define. 

-----

I think one thing is clear in my mind.. the name I like most for this feature is **Parameter** **Transformers**

![image](https://user-images.githubusercontent.com/1430011/44124610-f4241998-9fe2-11e8-9cb3-ccc645122f81.png)

**Comment by JamesNK** [comment]
> more like: we'd want to allow user code to plug in at the right places to take control.

Yea, I was using them as the main example.

> Maybe we're saying the same thing, but I think this needs to be done during the phase where we create the endpoints.

👍 

> I think it should replace the default casing. This is a stylistic choice for building a URL-space, and I think you should only do that with conviction.

👍 This is what Json.NET does as well

> I think user-code needs to decide.

Sure. I think area/controller/action is a good default though. Do those values have a name that could be applied to them for an option? `TransformMvcParameters { get; set; }`?

> I think defining how parameters get transformed has to be part of the route/template/endpoint definition rather than global.

What about a ParameterTransformer(tm) being in Endpoint metadata. EndpointBuilder could then have some method to set it if you desire. And MVC could have a global option to set it for all attribute routes and MapRoutes.

Or were you thinking something like per-parameter metadata?

**Comment by rynowak** [comment]
> Or were you thinking something like per-parameter metadata?

Hold on to your hats. We're about to go #MaximumRyan

I suggest we do the following.

Repurpose `Contraints` as `Policies` - what is a constraint really, but a specific type of policy. This means that you can specify policies inline in the template or out-of-line in the object literal when building a route pattern.

Example:
`/{controller=Home:snakecase}/{action=Index:snakecase}/{id?}`

OR

`/{controller=Home}/{action=Index}/{id?}`, constraints: new object{ controller = new SnakeCaseParameterTransformer() }

---

So, we define a new marker interface `IParameterPolicy` that `IRouteConstraint` inherits. Then we rename `MatchProcessorFactory` to `ParameterPolicyFactory`. 

We extend the set of things you can pass in a `constraints: new { ... }` literal to include anything that inherits from `IParameterPolicy` and `object[]` - today you can't use the object literal to specify multiple constraints (which is dumb). 

---

Then we can layer additional sugar at the MVC layer on top of this if we think it's valuable (I think the above is already good enough).

**Comment by JamesNK** [comment]
Yeah sounds ok to me. Specifying an inline transformer is cute. I think a global setting on MVC that automatically adds a transformer to area/controller/action is what people will use most.

**Comment by JamesNK** [comment]
I looked into this some more -
* The resolving of parameters happens almost entirely in MVC. Tokens are replaced in `AttributeRouteInfo`, e.g. `Blog/[action]/{postId?}` becomes `Blog/Edit/{postId?}`, in `ControllerActionDescriptorBuilder`
    Likewise the flattening of conventional routes happens in `MvcEndpointDataSource`, e.g. `{controller}\{action}\{id?}` becomes `Account\View\{id?}`
    If we add parameter transformers then some logic is going to leak out into the endpoint data source, or we provide some APIs to resolve the endpoint paths.
* `LinkGenerator` is still built around `RouteTemplate`, not `RoutePattern`, and so currently doesn't have access to policies. Should probably wait until `LinkGenerator` uses patterns.

**Comment by JamesNK** [comment]
Hi everyone

We have finished adding parameter transformers to routing and MVC. It will be available in 2.2 preview 3. A custom parameter transformer is used to change how parameters are match and generated.

Create a custom transformer:
```cs
    public class SlugifyParameterTransformer : IParameterTransformer
    {
        public string Transform(string value)
        {
            // Slugify value
            return Regex.Replace(value, "([a-z])([A-Z])", "$1-$2", RegexOptions.None, TimeSpan.FromMilliseconds(100)).ToLower();
        }
    }
```
Configure services:
```cs
            services
                .AddRouting(options =>
                {
                    options.ConstraintMap["slugify"] = typeof(SlugifyParameterTransformer);
                });
```
Configure route:
```cs
                routes.MapRoute(
                    "ConventionalTransformerRoute",
                    "ConventionalTransformerRoute/{controller:slugify}/{action=Index}/{param:slugify?}",
                    defaults: null,
                    constraints: new { controller = "ConventionalTransformer" });
```
Matched and generated URL changes from: `http://localhost/ConventionalTransformerRoute/ConventionalTransformer/Index`
To:
`http://localhost/ConventionalTransformerRoute/conventional-transformer/Index`

If you are using attribute routing you could either have the transformer defined in the route attribute:
```cs
    [Route("/{controller:slugify}")]
    public class EndpointRoutingController : Controller
    {
    }
```
Note that you can apply a parameter transformer to a non-area/controller/action parameter, and it will be transformed during link generation, but it is not automatically converted back to its initial value after matching.

If you are using attribute routing and you want all URLs transformed in your app then there is an API convention called `RouteTokenTransformerConvention` (@OlsonDev this is similar to what you are doing):
```cs
services
    .AddMvc(options =>
    {
        // Add route token transformer to all controllers
        options.Conventions.Add(new RouteTokenTransformerConvention(new SlugifyParameterTransformer()));
    })
    .SetCompatibilityVersion(CompatibilityVersion.Latest);
```

**Comment by JamesNK** [comment]
I'm going to close this as done. If you have feedback you can leave it here or create a new issue. If we identify changes we want to make before 2.2 goes live then we can reopen the issue.
